### ISSUE TITLE
BrickPi+ EV3motor, sensors python cannot connect 

### ISSUE THREAD

**Issue by jhconning** [rhempel/ev3dev-lang-python#282]
**System information:**
- Hardware
   - RaspberryPi 3 Model B
   - BrickPi+ (with battery pack on)
   - EV3 motors and sensors

- I installed ev3dev using image: ev3dev-jessie-rpi2-generic-2016-12-21.zip
- kernel:  4.4.32-17-ev3dev-rpi2
- My host computer is running: Windows10 
- I am connecting using: ethernet (SSH)
- ev3dev-lang-python version: python 3.4.2

My changes to config.txt (only enabled options):

    dtoverlay=brickpi
    init_uart_clock=32000000
    dtparam=brickpi_battery=okay
    dtoverlay=pi3-disable-bt

**My issue:**

I cannot control motors using ev3dev.brickpi methods even though I can control them via command line using examples from [Using the Tacho-Motor Class](http://www.ev3dev.org/docs/tutorials/tacho-motors/)

For example the following:

    import ev3dev.brickpi as ev3
    m = ev3.LargeMotor('outB')
    m.connected

produces output ```False```.  I get 'not connected' errors when I try other methods.

As mentioned, I *can* control the same LargeMotor on port B directly from the command line as described here http://www.ev3dev.org/docs/tutorials/tacho-motors/

The motors appear to be assigned like this:

    robot@ev3dev:~$ for f in /sys/class/tacho-motor/*; do echo -n "$f: "; cat $f/address; done
    /sys/class/tacho-motor/motor0: ttyAMA0:MA
    /sys/class/tacho-motor/motor1: ttyAMA0:MB
    /sys/class/tacho-motor/motor2: ttyAMA0:MC
    /sys/class/tacho-motor/motor3: ttyAMA0:MD

From reading issues #279 and #115 I suspect that because I'm on a BrickPi (no autodetection of sensors... and perhaps also motors)  I may have to somehow tell python what ports/devices to use manually.  Those discussions talk of using the LegoPort class but from the documentation and with my pretty limited understanding I cannot find a simple clear example to guide me on how to do just that for motors.  I have the motors and sensors that came with the retail Lego EV3 kit.  Any suggestions about how to proceed? Thanks.

**Comment by WasabiFan** [comment]
Replace this:
```python
m = ev3.LargeMotor('outB')
```

with this:
```python
m = ev3.LargeMotor(ev3.OUTPUT_B)
```

The ports are named `out{A,B,C,D}` on the EV3, but on BrickPi they have the names that you mentioned in your post: `ttyAMA0:M{A,B,C,D}`. By using the constants corresponding to each port, you can just change the import line to change the platform; you could also replace the hard-coded string.  Where did you get that sample code? We should update it to recommend the proper method.

**Comment by jhconning** [comment]
Thank you! That works.

You asked where I found the code.  That's the problem. There does not seem to be (please let me know if otherwise) a place in the documentation aimed at guiding somebody trying to make things work on the BrickPi.  I just adapted code from the README for this repo and [this page](https://sites.google.com/site/ev3python/learn_ev3_python/using-motors) from <ev3python.com>.

Actually, I hope it is not inappropriate that I have added 'sensors' to the issue title, to make this issue about improving the documentation for connecting the BrickPi to motors and sensors with python.  

It would be very useful to create a documentation page that illustrates how a BrickPi user might connect a motor and say a touchsensor in python.  The only clearly marked BrickPi section that I see in the docs is [step 3a in the Getting Started Guide](http://www.ev3dev.org/docs/getting-started/#step-3a-raspberry-pi-only-update-options-in-configtxt)  where changes to config.txt are explained.  

I would be happy to try to contribute to such a page, if I could finish trying to understand how to now also connect a sensor. 

From what you have told me and #280 I infer that if I were to try to connect an EV3 touch sensor to port 1 on a BrickPi, I would have to let the port know the type of sensor I am connecting.   From the [supported sensors table](http://www.ev3dev.org/docs/sensors/#supported-sensors) I see the LEGO EV3 Touch Sensor Driver(Module) is listed as "lego-ev3-touch (ev3-analog-sensor)" which suggests this code:

    import ev3dev.brickpi as ev3
    p = ev3.LegoPort(ev3.INPUT_1)
    p.mode ="ev3-analog"
    p.set_device = "lego-ev3-touch"
    
(Notice that I used "ev3-analog" rather than what I would have inferred from the sensor table which uses 'ev3-analog-sensor'  -- is that a typo or am I missing something?).  I feel the documentation could benefit with simple 'recipes' like this that explain the broad principles at play in layman's terms. 

Now that ev3 python knows what kind of sensor is attached I thought I should now be able to use the touch sensor like so:
  
    ts = ev3.TouchSensor(ev3.INPUT_1)
    print(ts.value())

This runs without error but seems to return a value of 1 whether I press the button or not.  Have I done something wrong? 

 As I said if I can make sense of this and others think it might be useful to have some simple working BrickPi connection examples in the documentation I could try my hand at writing something up.

**Comment by jhconning** [comment]
I'm closing the issue now because with the help @WasabiFan provided it's now clearer how to get things up and running with the BrickPi.  The changes he's proposed to the README should help other BrickPi users. 

After I've had a chance to play around some more to better understand the library (and sensors in particular) I may submit a page of documentation for the BrickPi (in the form of a PR) for consideration to supplement the documentation.  I'm closing this issue now since it may not be for another week or so..

What I have in mind a page that BrickPi users could consult that summarizes in one place the main issues/differences they will encounter and some solutions.  An (incomplete) list of points that might be included: 

- get right image for pi, pi2/pi3
- Must make changes to config.txt (this already is explained) 
- key hardware interface differences
    - no Brickman interface obviously
    - cannot expect to connect the pi to an HDMI screen and boot to a login screen. Must instead log in via SSH.
    - Easiest initial connection is probably via ethernet
           - link to instructions on how to connect pi to wifi from command prompt
           - similar link for Bluetooth connection
     - Brickpi cannot autodetect connected devices, must be set manually,
     - Dexter Industries [diagrams with input and output port positions ](https://www.dexterindustries.com/BrickPi/getting-started/attaching-lego/)
- Software differences
     - use `from ev3dev.BrickPi import *' not `from ev3dev.ev3 ...`
     - specify output port j as  OUTPUT_j, e.g.  `m = LargeMotor(OUTPUT_A)`
     - specify input sensors e.g. `TouchSensor(INPUT_1)`
          - how to find sensor information and  to properly set `mode` and `set_device`(?)
    - simple python code examples of the above.

**Comment by WasabiFan** [comment]
Sorry, it looks like I never responded directly to this thread. I opened #283 in response to your post, which includes the beginnings of this work. I was hoping to do exactly what you proposed here, so if you would be willing to do that I'd appreciate the help!

**Comment by dlech** [comment]
> key hardware interface differences

See http://www.ev3dev.org/docs/platform-comparison/
